### PR TITLE
Add tests for dashboard page

### PR DESCRIPTION
**Pull request by danascheider**
## Context

[**Add tests for dashboard page**](https://trello.com/c/IESUCSO2/79-add-tests-for-dashboard-page)

As part of the Jest epic, we needed to add tests covering the dashboard page. In the process, we uncovered some issues/edge cases in the context, having to do with state updates on unmounted components. Specifically, I found that the `logOutAndRedirect` function in the `AppProvider` was changing `cookies`, triggering the `fetchProfileData` function to run again when logging out from the dashboard or shopping list page. (The `fetchProfileData` function runs in a `useEffect` hook with `cookies` in its dependency array.)

## Changes

* Tests covering the dashboard page in cases where user is logged in, logged out, or has a token rejected by the server
* Code changes for better lifecycle management of the `AppContext`

## Considerations

Implementing these tests was slow going because of memory leaks and edge cases they uncovered in the app code. I've fixed things in a way to make the tests pass and the app work at the same time. Since we only have partial test coverage, it's important not to change more than necessary all at once.

## Manual Test Cases

Because changes were made to the `AppProvider`, comprehensive testing is required of any pages and components that use the `AppContext`.

### CODE REVIEW

File: src/contexts/appContext.js [hunk 1]
```diff
@@ -46,15 +47,31 @@ const AppProvider = ({ children, overrideValue = {} }) => {
...
     mountedRef.current = false
-  }, [mountedRef])
+  }, [])
 
   const displayFlash = useCallback((type, message, header = null) => {
     setFlashProps({ type, message, header })
     setFlashVisible(true)
     window.scrollTo(0, 0)
-  }, [setFlashProps, setFlashVisible])
+  }, [])
```
Comment (danascheider) [inline src/contexts/appContext.js:56]: Don't need to include state setters in dependency arrays.

File: src/contexts/appContext.js [hunk 2]
```diff
@@ -46,15 +47,31 @@ const AppProvider = ({ children, overrideValue = {} }) => {
   const setShouldRedirectTo = useCallback(path => {
     setRedirectPath(path)
     mountedRef.current = false
-  }, [mountedRef])
+  }, [])
```
Comment (danascheider) [inline src/contexts/appContext.js:50]: Don't need to include refs in dependency arrays.
